extra/man/upstart-udev-bridge.8
.TH upstart\-udev\-bridge 8 2011-03-08 upstart
.\"
.SH NAME
upstart\-udev\-bridge \- Bridge between Upstart and udev
.\"
.SH SYNOPSIS
.B upstart\-udev\-bridge
.RI [ OPTIONS ]...
.\"
.SH DESCRIPTION
.B upstart\-udev\-bridge
receives information about kernel uevents that
.BR udev (8)
has completed and creates
.BR init (8)
events for them.

It emits events which match the pattern "\fIS\fP\-device\-\fIA\fP" where
\(aqS\(aq is the udev \fIsubsystem\fP and \(aqA\(aq is one of
\fIadded\fR, \fIchanged\fR or \fIremoved\fR,
which correspond to the past tense of the udev \(aqaction\(aq.

See \fBudev\fP(7) and for further details.

Assuming \fI/sys\fP is mounted, possible values for \fIsubsystem\fP for
your system are viewable via \fI/sys/class/\fP.
.\"
.SH OPTIONS
.\"
.TP
.B \-\-daemon
Detach and run in the background.
.\"
.TP
.B \-\-debug
Enable debugging output.
.\"
.TP
.B \-\-help
Show brief usage summary.
.\"
.TP
.B \-\-no\-strip
Do not modify udev message contents. By default, all udev data will have
non-printable bytes removed. This option reverts the behaviour to not
modifying the data in any way. This option is not recommended since
specifying it
.I could
result in
.BR udev (7)
events not being emitted as Upstart events should any of your
hardware expose non-printable characters to the kernel. You can establish
whether this is happening by running
.I without
specifying this option
but with the 
.B \-\-debug
option.
.\"
.TP
.B \-\-verbose
Enable verbose output.
.\"
.SH EXAMPLES

Some examples of Upstart events generated by this bridge:
.IP net\-device\-added
Event emitted when a network device is added.
.IP net\-device\-removed
Event emitted when a network device is removed.
.IP graphics\-device\-added
Event emitted when a graphics device is available to the system.
.\"
.SH AUTHOR
Written by Scott James Remnant
.RB < scott@netsplit.com >
.\"
.SH BUGS
Report bugs at 
.RB < https://launchpad.net/ubuntu/+source/upstart/+bugs >
.\"
.SH COPYRIGHT
Copyright \(co 2009,2010,2011 Canonical Ltd.
.PP
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH SEE ALSO
.BR init (5)
.BR init (8)
.BR udev (7)


NEWS
1.13  xxxx-xx-xx ""

	* NEWS: Begin new release.

1.12.1  2014-03-11 "Standing Room Only"

	* Fixed re-exec breakage caused by incorrect D-Bus API usage.
	* Tests updated to catch this issue.

1.12  2014-03-07 "Water Table"

	* Stop Upstart indirectly stealing console ownership (most noticeably when
	  run in a container).
	* Fixed bug that could result in incorrect output from initctl(8)
	  environment commands.
	* Fixed incorrect variable usage which could result in a crash
	  (LP: #1222705).
	* Fixed stateful re-exec issue triggered by invalid jobs
	  (LP: #1269731).
	* Stateful re-exec fix to serialise the D-Bus session bus
	  (LP: #1258098).
	* init-checkconf(8) now uses a Session Init and can also now be
	  run as root.
	* upstart-socket-bridge(8) now supports IPv6 (thanks to Kai Mast and
	  Cameron Norman).
	* Allow telinit(8) to work on systems with no D-Bus system bus.
	* Numerous test improvements.
	* Lots of man-page improvements.

1.11  2013-11-14 "Quoi de neuf?"

	* Add '--no-dbus' command-line option for the init(8) daemon (allowing
	  event propagation to the Session Init via the upstart-event-bridge(8) to
	  be disabled).
	* Serialise Session Init job environment table on re-exec (LP: #1238078).
	* upstart-{dbus,socket}-bridge(8): Don't clobber existing PATH (LP: #1234898).
	* upstart-file-bridge(8):
	  - Allow watching for directory creation (LP: #1221466).
	  - Reduce default output.
	* upstart-local-bridge(8): Handle arbitrary numbers of pairs and additional
	  checks on input validity (LP: #1234898).
	* Improved Session Init shutdown speed (LP: #1227212).
	* New configure options: '--disable-local-bridge', '--disable-socket-bridge'.
	* Added upstart-file-bridge(8) and Session Init re-exec integration tests.
	* Retain umask for Session Init (LP: #1240686).
	* Make Session Init connect to D-Bus Session Bus on request
	  (LP: #1203595, #1235649)
	* Improved build-time ABI compliance checking.
	* Doc updates and lots of test refinements.

1.10  2013-08-23 "Swim to the buoy"

	* upstart-local-bridge: New bridge for starting jobs on local
	  socket connections.
	* upstart-dconf-bridge: New bridge for Session Inits.
	* upstart-dbus-bridge: New '--bus-name' option to allow bus name
	  variable to be included in dbus-event(7).
	* New "reload signal" stanza to allow jobs to specify a custom signal
	  that will be sent to the main process (rather than the default
	  SIGHUP).
	* Inclusion of Session Init sample jobs.
	* Re-exec fixes for handling chroot sessions.
	* Shutdown fix for Session Inits.
	* New python3 module and accompanying integration test suite for
	  testing Upstart running as PID 1 and as a Session Init
	  (priveleged and non-privileged).

1.9.1  2013-07-04 "The all-inclusive deal"

	* Fix respawn behaviour regression (introduced in version 1.8) affecting
	  jobs which also specify shell meta-characters in the exec stanza
	  (LP: #1181789).
	* Fix common bug which could cause a bridge to crash (LP: #1197225).
	* Various fixes for libupstart.
	* initctl2dot: Layout improvements.

1.9  2013-06-28 "This village isn't even on the map"

	* Apparmor support added by means of two new stanzas: 'apparmor load'
	  and 'apparmor switch' (thanks Marc Deslauriers).
	* Stateful re-exec: now serialises all objects, not just the minimum set.
	* Allow inherited environment variables to be unset for Session Inits.
	* Ability to specify multiple configuration directories when running as
	  a Session Init.
	* libupstart: Client library for communicating with Upstart.
	* upstart-dbus-bridge: New bridge to allow jobs to react to D-Bus signals.
	* The usual round of fixes and doc improvements.

1.8  2013-03-22 "Scone + Radish = Gherkin?"

	* upstart-file-bridge: New bridge to allow jobs to react to
	  file, directory and file glob events (create/modify/delete).
	* upstart-monitor: Simple cli/gui tool that shows event flows.

1.7  2013-03-04 "We've all got one now"

	* New initctl commands: set-env, unset-env, get-env, list-env,
	  reset-env, list-sessions (all except last with corresponding
	  D-Bus methods).
	* New D-Bus-only signals EventEmitted, Restarted, and EndSession method.
	* Ability to run with PID >1 to allow Upstart to manage a user session.
	  Running Upstart as a 'Session Init' in this way provides features
	  above and beyond those provided by the original User Jobs such that
	  the User Job facility has been removed entirely: to migrate from
	  a system using User Jobs, simply ensure the user session is started with
	  'init --user'.
	* New upstart-event-bridge bridge which proxies system-level events down
	  to Session Inits, allowing users jobs to react to udev events.
	* Ability to read job configuration and override files from multiple
	  freedesktop-compliant locations (Session Init only).
	* Ability to shutdown both via a system shutdown request and via a user
	  logout request (Session Init only).

1.6.1  2012-12-07 "It glows in the dark"

	* Improved re-exec performance.
	* Minor logger fixes for unflushed data.
	* Handle re-exec scenario when requested from within a chroot.
	* Minor serialisation data format change to handle chroots and
	  user sessions.
	* Added extra re-exec tests including explicit upgrade tests reading
	  from pre-prepared JSON data files.
	* Make jobs running within a chroot log their output within the chroot.
	* Added "Restart" D-Bus method.
	* Changed 'telinit u' to use "Restart" D-Bus method rather than
	  sending SIGTERM to play nicely when busybox(1) is init.
	* Added "GetState" D-Bus method allowing current serialised internal state
	  to be queried.

1.6  2012-11-15 "The turbine, the fish and the glass partition"

	* 'telinit u' now performs a stateful re-exec, allowing Upstart
	  to continue to supervise jobs after an upgrade of either
	  itself, or any of its dependent libraries. Adds dependency
	  on libjson (JSON-C).
	* Added initramfs-less boot support.
	* Allow tests to work in sbuild environment.
	* Improved error handling.

extra/upstart-udev-bridge.c
/* upstart
 *
 * Copyright © 2009 Canonical Ltd.
 * Author: Scott James Remnant <scott@netsplit.com>.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif /* HAVE_CONFIG_H */


#include <libudev.h>

#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <ctype.h>

#include <nih/macros.h>
#include <nih/alloc.h>
#include <nih/string.h>
#include <nih/io.h>
#include <nih/option.h>
#include <nih/main.h>
#include <nih/logging.h>
#include <nih/error.h>

#include <nih-dbus/dbus_connection.h>
#include <nih-dbus/dbus_proxy.h>

#include "dbus/upstart.h"
#include "com.ubuntu.Upstart.h"


/* Prototypes for static functions */
static void udev_monitor_watcher (struct udev_monitor *udev_monitor,
				  NihIoWatch *watch, NihIoEvents events);
static void upstart_disconnected (DBusConnection *connection);
static void emit_event_error     (void *data, NihDBusMessage *message);

static char *make_safe_string    (const void *parent, const char *original);

/**
 * daemonise:
 *
 * Set to TRUE if we should become a daemon, rather than just running
 * in the foreground.
 **/
static int daemonise = FALSE;

/**
 * upstart:
 *
 * Proxy to Upstart daemon.
 **/
static NihDBusProxy *upstart = NULL;

/**
 * no_strip_udev_data:
 *
 * If TRUE, do not modify any udev message data (old behaviour).
 * If FALSE, use make_safe_string () to cleanse udev strings.
 **/
static int no_strip_udev_data = FALSE;

/**
 * options:
 *
 * Command-line options accepted by this program.
 **/
static NihOption options[] = {
	{ 0, "daemon", N_("Detach and run in the background"),
	  NULL, NULL, &daemonise, NULL },
	{ 0, "no-strip", N_("Do not strip non-printable bytes from udev message data"),
	  NULL, NULL, &no_strip_udev_data, NULL },

	NIH_OPTION_LAST
};


int
main (int   argc,
      char *argv[])
{
	char **              args;
	DBusConnection *     connection;
	struct udev *        udev;
	struct udev_monitor *udev_monitor;
	int                  ret;

	nih_main_init (argv[0]);

	nih_option_set_synopsis (_("Bridge udev events into upstart"));
	nih_option_set_help (
		_("By default, upstart-udev-bridge does not detach from the "
		  "console and remains in the foreground.  Use the --daemon "
		  "option to have it detach."));

	args = nih_option_parser (NULL, argc, argv, options, FALSE);
	if (! args)
		exit (1);

	/* Initialise the connection to Upstart */
	connection = NIH_SHOULD (nih_dbus_connect (DBUS_ADDRESS_UPSTART, upstart_disconnected));
	if (! connection) {
		NihError *err;

		err = nih_error_get ();
		nih_fatal ("%s: %s", _("Could not connect to Upstart"),
			   err->message);
		nih_free (err);

		exit (1);
	}

	upstart = NIH_SHOULD (nih_dbus_proxy_new (NULL, connection,
						  NULL, DBUS_PATH_UPSTART,
						  NULL, NULL));
	if (! upstart) {
		NihError *err;

		err = nih_error_get ();
		nih_fatal ("%s: %s", _("Could not create Upstart proxy"),
			   err->message);
		nih_free (err);

		exit (1);
	}

	/* Initialise the connection to udev */
	nih_assert (udev = udev_new ());
	nih_assert (udev_monitor = udev_monitor_new_from_netlink (udev, "udev"));
	nih_assert (udev_monitor_enable_receiving (udev_monitor) == 0);
	udev_monitor_set_receive_buffer_size(udev_monitor, 128*1024*1024);

	NIH_MUST (nih_io_add_watch (NULL, udev_monitor_get_fd (udev_monitor),
				    NIH_IO_READ,
				    (NihIoWatcher)udev_monitor_watcher,
				    udev_monitor));

	/* Become daemon */
	if (daemonise) {
		if (nih_main_daemonise () < 0) {
			NihError *err;

			err = nih_error_get ();
			nih_fatal ("%s: %s", _("Unable to become daemon"),
				   err->message);
			nih_free (err);

			exit (1);
		}

		/* Send all logging output to syslog */
		openlog (program_name, LOG_PID, LOG_DAEMON);
		nih_log_set_logger (nih_logger_syslog);
	}

	/* Handle TERM and INT signals gracefully */
	nih_signal_set_handler (SIGTERM, nih_signal_handler);
	NIH_MUST (nih_signal_add_handler (NULL, SIGTERM, nih_main_term_signal, NULL));

	if (! daemonise) {
		nih_signal_set_handler (SIGINT, nih_signal_handler);
		NIH_MUST (nih_signal_add_handler (NULL, SIGINT, nih_main_term_signal, NULL));
	}

	ret = nih_main_loop ();

	return ret;
}


static void
udev_monitor_watcher (struct udev_monitor *udev_monitor,
		      NihIoWatch *         watch,
		      NihIoEvents          events)
{
	struct udev_device *    udev_device;
	nih_local char *        subsystem = NULL;
	nih_local char *        action = NULL;
	nih_local char *        kernel = NULL;
	nih_local char *        devpath = NULL;
	nih_local char *        devname = NULL;
	nih_local char *        name = NULL;
	nih_local char **       env = NULL;
	const char *            value = NULL;
	size_t                  env_len = 0;
	DBusPendingCall *       pending_call;
	char                 *(*copy_string)(const void *, const char *) = NULL;


	udev_device = udev_monitor_receive_device (udev_monitor);
	if (! udev_device)
		return;

	copy_string = no_strip_udev_data ? nih_strdup : make_safe_string;

	value = udev_device_get_subsystem (udev_device);
	subsystem = value ? copy_string (NULL, value) : NULL;

	value = udev_device_get_action (udev_device);
	action = value ? copy_string (NULL, value) : NULL;

	value = udev_device_get_sysname (udev_device);
	kernel = value ? copy_string (NULL, value) : NULL;

	value = udev_device_get_devpath (udev_device);
	devpath = value ? copy_string (NULL, value) : NULL;

	value = udev_device_get_devnode (udev_device);
	devname = value ? copy_string (NULL, value) : NULL;

	/* Protect against the "impossible" */
	if (! action)
		goto out;

	if (! strcmp (action, "add")) {
		name = NIH_MUST (nih_sprintf (NULL, "%s-device-added",
					      subsystem));
	} else if (! strcmp (action, "change")) {
		name = NIH_MUST (nih_sprintf (NULL, "%s-device-changed",
					      subsystem));
	} else if (! strcmp (action, "remove")) {
		name = NIH_MUST (nih_sprintf (NULL, "%s-device-removed",
					      subsystem));
	} else {
		name = NIH_MUST (nih_sprintf (NULL, "%s-device-%s",
					      subsystem, action));
	}

	env = NIH_MUST (nih_str_array_new (NULL));

	if (kernel) {
		nih_local char *var = NULL;

		var = NIH_MUST (nih_sprintf (NULL, "KERNEL=%s", kernel));
		NIH_MUST (nih_str_array_addp (&env, NULL, &env_len, var));
	}

	if (devpath) {
		nih_local char *var = NULL;

		var = NIH_MUST (nih_sprintf (NULL, "DEVPATH=%s", devpath));
		NIH_MUST (nih_str_array_addp (&env, NULL, &env_len, var));
	}

	if (devname) {
		nih_local char *var = NULL;

		var = NIH_MUST (nih_sprintf (NULL, "DEVNAME=%s", devname));
		NIH_MUST (nih_str_array_addp (&env, NULL, &env_len, var));
	}

	if (subsystem) {
		nih_local char *var = NULL;

		var = NIH_MUST (nih_sprintf (NULL, "SUBSYSTEM=%s", subsystem));
		NIH_MUST (nih_str_array_addp (&env, NULL, &env_len, var));
	}

	if (action) {
		nih_local char *var = NULL;

		var = NIH_MUST (nih_sprintf (NULL, "ACTION=%s", action));
		NIH_MUST (nih_str_array_addp (&env, NULL, &env_len, var));
	}

	for (struct udev_list_entry *list_entry = udev_device_get_properties_list_entry (udev_device);
	     list_entry != NULL;
	     list_entry = udev_list_entry_get_next (list_entry)) {
		nih_local char *udev_name = NULL;
		nih_local char *udev_value = NULL;
		nih_local char *var = NULL;

		udev_name = copy_string (NULL, udev_list_entry_get_name (list_entry));

		if (! strcmp (udev_name, "DEVPATH"))
			continue;
		if (! strcmp (udev_name, "DEVNAME"))
			continue;
		if (! strcmp (udev_name, "SUBSYSTEM"))
			continue;
		if (! strcmp (udev_name, "ACTION"))
			continue;

		udev_value = copy_string (NULL, udev_list_entry_get_value (list_entry));

		var = NIH_MUST (nih_sprintf (NULL, "%s=%s", udev_name, udev_value));
		NIH_MUST (nih_str_array_addp (&env, NULL, &env_len, var));
	}

	nih_debug ("%s %s", name, devname ? devname : "");

	pending_call = upstart_emit_event (upstart,
			name, env, FALSE,
			NULL, emit_event_error, NULL,
			NIH_DBUS_TIMEOUT_NEVER);

	if (! pending_call) {
		NihError *err;
		int saved = errno;

		err = nih_error_get ();
		nih_warn ("%s", err->message);

		if (saved != ENOMEM && subsystem)
			nih_warn ("Likely that udev '%s' event contains binary garbage", subsystem);

		nih_free (err);
	}

	dbus_pending_call_unref (pending_call);

out:
	udev_device_unref (udev_device);
}


static void
upstart_disconnected (DBusConnection *connection)
{
	nih_fatal (_("Disconnected from Upstart"));
	nih_main_loop_exit (1);
}

static void
emit_event_error (void *          data,
		  NihDBusMessage *message)
{
	NihError *err;

	err = nih_error_get ();
	nih_warn ("%s", err->message);
	nih_free (err);
}

/**
 * make_safe_string:
 * @parent: parent,
 * @original: original string.
 *
 * Strip non-printable and non-blank bytes from specified string.
 *
 * Notes:
 *
 * Sadly, this is necessary as some hardware (such as battery devices)
 * exposes non-printable bytes in their descriptive registers to the
 * kernel. Since neither the kernel nor udev specify any encoding for
 * udev messages, these (probably bogus) bytes get passed up to userland
 * to deal with. This is sub-optimal since it implies that _every_
 * application that processes udev messages must perform its own
 * sanitizing on the messages. Let's just hope they all deal with the
 * problem in the same way...
 *
 * Note that *iff* the kernel/udev did specify an encoding model, this
 * problem could go away since one of the lower layers could then
 * detect the out-of-bound data and deal with it at source. All instances
 * of this issue seen so far seem to indicate the binary control data
 * being presented by the hardware is in fact bogus ("corruption") and
 * looks like some block of memory has not been initialized correctly.
 *
 * The approach taken here is to simulate the approach already adopted
 * by 'upower' (up_device_supply_make_safe_string()), with the exception
 * that we also allow blank characters (such as tabs).
 *
 * Returns a copy of @original stripped of all non-printable and
 * non-blank characters, or NULL if insufficient memory.
 **/
char *
make_safe_string (const void *parent, const char *original)
{
	size_t   len;
	size_t   i, j;
	char    *cleaned;

	nih_assert (original);

	len = strlen (original);

	cleaned = nih_alloc (parent, len + 1);

	if (! cleaned)
		return NULL;

	for (i=0, j=0; i < len; ) {
		/* Skip over bogus bytes */
		if (! (isprint (original[i]) || isblank (original[i]))) {
			i++;
			continue;
		}

		/* Copy what remains */
		cleaned[j] = original[i];
		i++; j++;
	}

	/* Terminate */
	cleaned[j] = '\0';

	if (i != j)
		nih_debug ("removed unexpected bytes from udev message data");

	/* Note that strictly we should realloc the string if
	 * bogus bytes were found (since it will now be shorter).
	 * However, since all the strings are short (and short-lived) we
	 * do not do this to avoid the associated overhead.
	 */
	return cleaned;
}
